src/services/utilities.ts

```diff
@@ -1158,6 +1158,35 @@ namespace ts {
         }
     }
 
+    /**
+     * Psuedo-binary searches the list of children for the first element whose end position exceeds the input position
+     * Returns `undefined` if no element satisfies the condition, or the index of the satisfying element otherwise.
+     * This could probably be written using our `binarySeach` helper, but the `keyComparator` would be hairy enough
+     * that it's probably worth just writing out the algorithm in full.
+     * @param {number} start - Start search index, inclusive
+     * @param {number} end - End search index, inclusive
+     */
+    function searchChildrenSlice(position: number, children: Node[], start = 0, end = children.length - 1): number | undefined {
+        const pivot = Math.floor((end - start)/2) + start;
+        if (!children[pivot]) {
+            return undefined;
+        }
+        if (position < children[pivot].end) {
+            // first element whose end position is greater than the input position
+            if (!children[pivot - 1] || position >= children[pivot - 1].end) {
+                return pivot;
+            }
+            if (pivot === start) {

```

Comment:
Should we do this check before we start subtracting things from `pivot`?

Reply:
The edge-of-array case (`!children[pivot - 1]`) differs in that the single element under consideration is defacto the first element satisfying the condition, so needs to return the element, so it needs to come before the "bail and return `undefined` because there's only one element in the set under inspection and it doesn't match" case here.

In short: The prior condition handles it.

Reply:
I can't tell how we know `pivot - 1 >= start`.